tools/rpkg/R/backend-dbplyr__duckdb_connection.R

```diff
@@ -0,0 +1,383 @@
+      bitwShiftR = function(a, b) sql_expr((CAST((!!a) %AS% INTEGER)) %>>% (CAST((!!b) %AS% INTEGER))),
+      log = function(x, base = exp(1)) {
+        if (isTRUE(all.equal(base, exp(1)))) {
+          sql_expr(LN(!!x))
+        } else {
+          sql_expr(LOG(!!x) / LOG(!!base))
+        }
+      },
+      log10 = sql_prefix("LOG10", 1),
+      log2 = sql_prefix("LOG2", 1),
+
+      # Function 	    Inf 	–Inf 	NaN 	NA
+      # is.finite() 	  FALSE FALSE FALSE FALSE
+      # is.infinite() 	TRUE 	TRUE 	FALSE FALSE
+      # is.nan() 	    FALSE FALSE TRUE 	FALSE
+      # is.na() 	      FALSE FALSE TRUE 	TRUE
+      is.na = function(a) build_sql("(", a, " IS NULL OR PRINTF('%f',", a, ") = 'nan')"),
+      is.nan = function(a) build_sql("(", a, " IS NOT NULL AND PRINTF('%f',", a, ") = 'nan')"),
+      is.infinite = function(a) build_sql("(", a, " IS NOT NULL AND REGEXP_MATCHES(PRINTF('%f',", a, "),'inf'))"),
+      is.finite = function(a) build_sql("(NOT (", a, " IS NULL OR REGEXP_MATCHES(PRINTF('%f',", a, "),'inf|nan')))"),
+      grepl = duckdb_grepl,
+
+      # Return índex where first match starts, -1 if not found
+      regexpr = function(p, x) {
+        build_sql(
+          "(LENGTH(LIST_EXTRACT(STRING_SPLIT_REGEX(", x, ", ", p,
+          "),0))-(LENGTH(LIST_EXTRACT(STRING_SPLIT_REGEX(", x, ", ", p,
+          "),0))+2)/(LENGTH(", x, ")+2)*(LENGTH(", x, ")+2)+1)"
+        )
+      },
+      round = function(x, digits) sql_expr(ROUND(!!x, CAST(ROUND((!!digits), 0L) %AS% INTEGER))),
+      as.Date = sql_cast("DATE"),
+      as.POSIXct = sql_cast("TIMESTAMP"),
+
+      # lubridate functions
+
+      month = function(x, label = FALSE, abbr = TRUE) {
+        if (!label) {
+          sql_expr(EXTRACT(MONTH %FROM% !!x))
+        } else {
+          if (abbr) {
+            sql_expr(STRFTIME(!!x, "%b"))
+          } else {
+            sql_expr(STRFTIME(!!x, "%B"))
+          }
+        }
+      },
+      quarter = function(x, type = "quarter", fiscal_start = 1, with_year = identical(type, "year.quarter")) {
+        if (fiscal_start != 1) {
+          stop("`fiscal_start` is not yet supported in DuckDB translation. Must be 1.", call. = FALSE)
+        }
+        if (is.logical(type)) {
+          type <- if (type) {
+            "year.quarter"
+          } else {
+            "quarter"
+          }
+        }
+        if (with_year == TRUE) {
+          type <- "year.quarter"
+        }
+        switch(type,
+          quarter = {
+            sql_expr(EXTRACT(QUARTER %FROM% !!x))
+          },
+          year.quarter = {
+            sql_expr((EXTRACT(YEAR %FROM% !!x) || "." || EXTRACT(QUARTER %FROM% !!x)))
+          },
+          date_first = {
+            sql_expr((CAST(DATE_TRUNC("QUARTER", !!x) %AS% DATE)))
+          },
+          date_last = {
+            build_sql("(CAST((DATE_TRUNC('QUARTER',", x, ") + INTERVAL '1 QUARTER' - INTERVAL '1 DAY') AS DATE))")
```

Comment:
I had some difficulties to use `sql_expr()` in all occasions and therefore used `build_sql()` if that seemed to be easier. So it would be more work without clear benefits (other than consistency). Or is it somehow unsafe to use `build_sql()`?

I think that sql_expr() with quite non-R - non-SQL approach is even more confusing and also requires `globalvariables` to avoid R-check NOTEs. I think that there also were some `NAMESPACE` issues as some %%-addition conflicted with reserved name in some package and R check insisted to import from that package even though the package was not even used.

Reply:
This particular `build_sql` was easy to change to `sql_expr` based on your example above, so I changed it to increase consistency. The `build_sql`s elsewhere I left as they were. The "namespace issue" was probably related to the use of `%||%` or `%:%` and maybe something that `R CMD check` complained about. I can't replicate the problem now, so it could also have been something session specific.